components/httpd_server/src/httpd/request.cpp

```diff
@@ -0,0 +1,202 @@
+  }
+
+  std::string Request::uri()
+  {
+    return uri_;
+  }
+
+  std::string Request::header(std::string key)
+  {
+    auto item = headers_.find(key);
+    if(item == headers_.end())
+    {
+      return "";
+    }
+    else
+    {
+      return item->second;
+    }
+  }
+
+  std::string Request::body()
+  {
+    return body_;
+  }
+
+  char* Request::parseMethod(char* cursor)
+  {
+    // Move to the next space...
+    char* end = cursor;
+    int count = 0;
+    while(*end && *end != ' ')
+    {
+      end++;
+      count++;
+    }
+
+    method_ = method::OTHER;
+
+    if (count == 3)
+    {
+      if      (0 == std::strncmp(cursor, "GET", 3)) method_ = method::GET;
+      else if (0 == std::strncmp(cursor, "PUT", 3)) method_ = method::PUT;
+    }
+    else if (count == 4)
+    {
+      if      (0 == std::strncmp(cursor, "POST", 4)) method_ = method::POST;
+      else if (0 == std::strncmp(cursor, "HEAD", 4)) method_ = method::HEAD;
+    }
+    else if (count == 5)
+    {
+      if      (0 == std::strncmp(cursor, "TRACE", 5)) method_ = method::TRACE;
+    }
+    else if (count == 6)
+    {
+      if      (0 == std::strncmp(cursor, "DELETE", 6)) method_ = method::DELETE;
+    }
+    else if (count == 7)
+    {
+      if      (0 == std::strncmp(cursor, "CONNECT", 7)) method_ = method::CONNECT;
+      else if (0 == std::strncmp(cursor, "OPTIONS", 7)) method_ = method::OPTIONS;
+    }
+
+    return end;
+  }
+
+  char* Request::parseUri(char* data)
+  {
+    // first remove trailing spaces
+    char *begin = data;
+    while (*begin && *begin == ' ') begin++;
+
+    char* end = begin;
+    while(*end && *end != ' ') end++;
+
+    uri_ = std::string(begin, end);
+
+    return end;
+  }
+
+  char* Request::parseHeaders(char* data)
+  {
+    char* var = data;
+
+    while(*var)
+    {
+      char* begin = var;
+
+      // Check where the end of line is. If at the beginning,
+      // We are at the end of the headers.
+      char* endofline = std::strstr(begin, "\r\n");
+      if (!endofline)
+      {
+        // This is incorrect!
+        break;
+      }
+      else if (endofline == begin)
+      {
+        var += 2;
+        break;
+      }
+
+      char* colon = std::strstr(begin, ":");
+
+      std::string variable = std::string(begin, colon);
+      // Move the cursor to the start of the next word
+      while (*colon && (*colon == ':' || *colon == ' ' )) colon++;
+
+      std::string value = std::string(colon, endofline);
+      headers_.insert({variable, value});
+      var = endofline + 2;
+    }
+
+    return var;
+  }
+
+  void Request::log()
+  {
+    ESP_LOGI("Request", "Method: %s", Request::toString(method()).c_str());
+    ESP_LOGI("Request", "URI   : %s", uri().c_str());
+    for(auto &h : headers_)
+    {
+      ESP_LOGI("Request", "h: '%s': '%s'", h.first.c_str(), h.second.c_str());
+    }
+    ESP_LOGI("Request", "body  : %s", body().c_str());
+  }
+
+  std::string Request::toString(method::eHTTPMethod_t code)
```

Comment:
If you convert a Request to string, I'd expect more than the string version of the method. Can you rename the function to `methodToString` or something similar?